### PR TITLE
Feature/559

### PR DESCRIPTION
**Pull request by OlegGordiichuk**
- Code refactoring
- Code cleaning
- Code optimization
- Add TrustOperation

### CODE REVIEW

File: Trust/Transactions/ViewControllers/TransactionsViewController.swift [hunk 1]
```diff
@@ -98,16 +112,44 @@ class TransactionsViewController: UIViewController {
...
-        super.viewDidAppear(animated)
+    private func tokensObservation() {
+        viewModel.setTransactionsObservation { [weak self] (changes: RealmCollectionChange) in
+            guard let strongSelf = self else { return }
+            let tableView = strongSelf.tableView
+            switch changes {
+            case .initial:
+                tableView.reloadData()
+                self?.endLoading()
+            case .update(_, let deletions, let insertions, let modifications):
```
Comment (vikmeup) [inline Trust/Transactions/ViewControllers/TransactionsViewController.swift]: this code looks repetitive with tokens as well, can we combine it?

File: Trust/Transactions/ViewControllers/TransactionsViewController.swift [hunk 2]
```diff
@@ -132,77 +174,60 @@ class TransactionsViewController: UIViewController {
...
+
+    deinit {
+        NotificationCenter.default.removeObserver(self)
     }
 }
 
 extension TransactionsViewController: StatefulViewController {
     func hasContent() -> Bool {
-        return viewModel.numberOfSections > 0
+        return !viewModel.transactions.isEmpty
```
Comment (vikmeup) [inline Trust/Transactions/ViewControllers/TransactionsViewController.swift]: can you add `hasContent ` to the `viewModel` itself?

File: Trust/Transactions/ViewControllers/TransactionsViewController.swift [hunk 3]
```diff
@@ -132,77 +174,60 @@ class TransactionsViewController: UIViewController {
...
     }
 }
 
 extension TransactionsViewController: UITableViewDelegate {
     func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
         tableView.deselectRow(at: indexPath, animated: true )
+        searchController.isActive = false
```
Comment (vikmeup) [inline Trust/Transactions/ViewControllers/TransactionsViewController.swift]: I also noticed that search controller has memory issues in the past, did you see the same?

File: Trust/Transactions/ViewControllers/TransactionsViewController.swift [hunk 4]
```diff
@@ -132,77 +174,60 @@ class TransactionsViewController: UIViewController {
...
-        let transaction = viewModel.item(for: indexPath.row, section: indexPath.section)
-        let cell = TransactionViewCell(style: .default, reuseIdentifier: TransactionViewCell.identifier)
-        cell.configure(viewModel: .init(
-                transaction: transaction,
-                config: session.config,
-                chainState: session.chainState,
-                currentWallet: session.account
-            )
-        )
+        let cell = tableView.dequeueReusableCell(withIdentifier: TransactionViewCell.identifier, for: indexPath) as! TransactionViewCell
```
Comment (vikmeup) [inline Trust/Transactions/ViewControllers/TransactionsViewController.swift]: Can you use `R` for an identifier?

File: Trust/Transactions/ViewModels/TransactionsViewModel.swift [hunk 5]
```diff
@@ -80,9 +98,90 @@ struct TransactionsViewModel {
         return value
     }
 
-    var isBuyActionAvailable: Bool {
-        switch config.server {
-        case .main, .kovan, .classic, .callisto, .ropsten, .rinkeby, .poa, .sokol, .custom: return false
+    func hederView(for section: Int) -> UIView {
```
Comment (vikmeup) [inline Trust/Transactions/ViewModels/TransactionsViewModel.swift]: Do you think view model should be handling `UIView`?

File: Trust/Transactions/ViewModels/TransactionsViewModel.swift [hunk 6]
```diff
@@ -80,9 +98,90 @@ struct TransactionsViewModel {
...
+        let conteiner = UIView()
+        conteiner.backgroundColor = self.headerBackgroundColor
+        let title = UILabel()
+        title.text = self.titleForHeader(in: section)
+        title.sizeToFit()
+        title.textColor = self.headerTitleTextColor
+        title.font = self.headerTitleFont
+        conteiner.addSubview(title)
+        title.translatesAutoresizingMaskIntoConstraints = false
+        let horConstraint = NSLayoutConstraint(item: title, attribute: .centerX, relatedBy: .equal, toItem: conteiner, attribute: .centerX, multiplier: 1.0, constant: 0.0)
```
Comment (vikmeup) [inline Trust/Transactions/ViewModels/TransactionsViewModel.swift]: is there a way to use iOS 9 auto layout API?

File: Trust/Transactions/Storage/TransactionsStorage.swift [hunk 7]
```diff
@@ -68,18 +75,33 @@ class TransactionsStorage {
         return tokens
     }
 
+    private func transactionCategory(for transactions: [Transaction]) -> [TransactionCategory] {
```
Comment (vikmeup) [inline Trust/Transactions/Storage/TransactionsStorage.swift]: If you change locale does it change `title` as well?

**Comment by vikmeup** [comment]
@OlegGordiichuk nice work! Looks much cleaner now, just some questions

File: Trust/Transactions/ViewControllers/TransactionsViewController.swift (hunk 1 above)
Comment (OlegGordiichuk) [reply Trust/Transactions/ViewControllers/TransactionsViewController.swift]: This issue will be in separate ticket with update of the transactions models.

File: Trust/Transactions/ViewControllers/TransactionsViewController.swift (hunk 2 above)
Comment (OlegGordiichuk) [reply Trust/Transactions/ViewControllers/TransactionsViewController.swift]: Done

File: Trust/Transactions/ViewControllers/TransactionsViewController.swift (hunk 3 above)
Comment (OlegGordiichuk) [reply Trust/Transactions/ViewControllers/TransactionsViewController.swift]: Hmm in transaction case there is no issue.

File: Trust/Transactions/ViewModels/TransactionsViewModel.swift (hunk 5 above)
Comment (OlegGordiichuk) [reply Trust/Transactions/ViewModels/TransactionsViewModel.swift]: Nice question but I think we should remove all additional logic from vc.

File: Trust/Transactions/ViewModels/TransactionsViewModel.swift (hunk 6 above)
Comment (OlegGordiichuk) [reply Trust/Transactions/ViewModels/TransactionsViewModel.swift]: Done

File: Trust/Transactions/ViewControllers/TransactionsViewController.swift (hunk 4 above)
Comment (OlegGordiichuk) [reply Trust/Transactions/ViewControllers/TransactionsViewController.swift]: Done

File: Trust/Transactions/Storage/TransactionsStorage.swift (hunk 7 above)
Comment (OlegGordiichuk) [reply Trust/Transactions/Storage/TransactionsStorage.swift]: Done

**Comment by vikmeup** [comment]
Nice work danger :) haha

**Comment by MillerApps** [comment]
# A message from Danger: "Thank you!"
image

**Comment by MillerApps** [comment]
You can ignore the warning: Delegate protocols should be class-only so they can be weakly referenced. It seems to be an isue the swiftLint gem using the wrong .swiftlint.yml. I'll Get it fixed.

File: Trust/Tokens/ViewControllers/NonFungibleTokensViewController.swift [hunk 8]
```diff
@@ -113,10 +113,11 @@ class NonFungibleTokensViewController: UIViewController {
         title.font = viewModel.headerTitleFont
         conteiner.addSubview(title)
         title.translatesAutoresizingMaskIntoConstraints = false
-        let horConstraint = NSLayoutConstraint(item: title, attribute: .centerX, relatedBy: .equal, toItem: conteiner, attribute: .centerX, multiplier: 1.0, constant: 0.0)
-        let verConstraint = NSLayoutConstraint(item: title, attribute: .centerY, relatedBy: .equal, toItem: conteiner, attribute: .centerY, multiplier: 1.0, constant: 0.0)
-        let leftConstraint = NSLayoutConstraint(item: title, attribute: .left, relatedBy: .equal, toItem: conteiner, attribute: .left, multiplier: 1.0, constant: 20.0)
-        conteiner.addConstraints([horConstraint, verConstraint, leftConstraint])
+        NSLayoutConstraint.activate([
+            title.centerXAnchor.constraint(equalTo: conteiner.centerXAnchor, constant: 0.0),
```
Comment (vikmeup) [inline Trust/Tokens/ViewControllers/NonFungibleTokensViewController.swift]: `conteiner ` => `container`

File: Trust/Transactions/ViewModels/TransactionsViewModel.swift [hunk 9]
```diff
@@ -87,8 +95,9 @@ struct TransactionsViewModel {
     }
 
     func titleForHeader(in section: Int) -> String {
-        let value = transactions[section].title
-        let date = TransactionsViewModel.formatter.date(from: value)!
+        let stringDate = transactions[section].title
+        let date = TransactionsViewModel.convert(stringDate)!
```
Comment (vikmeup) [inline Trust/Transactions/ViewModels/TransactionsViewModel.swift]: can we avoid force unwrap?

File: Trust/Transactions/ViewModels/TransactionsViewModel.swift [hunk 10]
```diff
@@ -184,4 +198,17 @@ struct TransactionsViewModel {
...
+
+    static func convert(_ date: String) -> Date? {
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = realmDateFormat
+        let date = dateFormatter.date(from: date)
+        return date
+    }
+
+    static func title(from date: Date) -> String {
+        let dateFormatter = DateFormatter()
```
Comment (vikmeup) [inline Trust/Transactions/ViewModels/TransactionsViewModel.swift]: can you make DateFormatter static? for the performance

File: Trust/Transactions/ViewModels/TransactionsViewModel.swift [hunk 11]
```diff
@@ -184,4 +198,17 @@ struct TransactionsViewModel {
         let predicate = NSCompoundPredicate(orPredicateWithSubpredicates: subpredicates)
         self.transactions = storage.transactionsCategory.filter(predicate)
     }
+
+    static func convert(_ date: String) -> Date? {
+        let dateFormatter = DateFormatter()
```
Comment (vikmeup) [inline Trust/Transactions/ViewModels/TransactionsViewModel.swift]: same here

File: Trust/Tokens/ViewControllers/NonFungibleTokensViewController.swift (hunk 8 above)
Comment (OlegGordiichuk) [reply Trust/Tokens/ViewControllers/NonFungibleTokensViewController.swift]: Done

File: Trust/Transactions/ViewModels/TransactionsViewModel.swift (hunk 10 above)
Comment (OlegGordiichuk) [reply Trust/Transactions/ViewModels/TransactionsViewModel.swift]: Done

File: Trust/Transactions/ViewModels/TransactionsViewModel.swift (hunk 11 above)
Comment (OlegGordiichuk) [reply Trust/Transactions/ViewModels/TransactionsViewModel.swift]: Done

**Pull request merged**